TYPESCRIPT.md

```diff
@@ -0,0 +1,196 @@
+## Typescriptifying Kibana Tips
+
+### Converting existing code
+
+To convert existing code over to typescript:
+1. rename the file from `.js` to either `.ts` (if there is no html or jsx in the file) or `.tsx` (if there is).
+2. Ensure tslint is running and installed in the IDE of your choice.  There will usually be some linter errors after the file rename.
+3. Auto-fix what you can. This will save you a lot of time! VSCode can be set to auto fix tslint errors when files are saved.
+
+### How to fix common typescript errors
+
+The first thing that will probably happen when you convert a `.js` file in our system to `.ts` is that some imports will be lacking types. 
+
+#### EUI component is missing types
+
+1. Check https://github.com/elastic/eui/issues/256 to see if they know it’s missing, if it’s not on there, add it.
+2. Temporarily get around the issue by using a declared module and exporting the missing types with the most basic types available. Bonus points if you write a PR yourself to the EUI repo to add the types, but having them available back in Kibana will take some time, as a new EUI release will need to be generated, then that new release pointed to in Kibana.  Best, to make forward progress, to do a temporary workaround.
+
+```ts
+declare module '@elastic/eui' {
+  export const EuiPopoverTitle: React.SFC<any>;
+}
+
+import { EuiPopoverTitle } from '@elastic/eui';
+```
+
+Note: I think this has the possibility to cause a conflict if you import another file which also declares the `@elastic/eui` module.  TODO: document what the best thing is to do if you hit this situation.
```

Comment:
At this point it might be useful to explain the difference between module declaration and augmentation, e.g.

In TypeScript module declaration can not be merged, which means each module can only be declared once. But it is possible to augment previously declared modules. The documentation about the distinction between module declaration and augmentation is sparse. The observed rules for `declare module '...' {}` in a `.d.ts` file seem to be:

* it is treated as a module declaration when the file itself is not a module
* it is treated as a module augmentation when the file itself is module

A `.d.ts` file is treated as a module if it contains any top-level `import` or `export` statements. That means that in order to write a module declaration the `import`s must be contained within the `declare` block and none must be located on the top level. Conversely, to write a module augmentation there must be at least one top-level `import` or `export` and the `declare` block must not contain any `import` statements.

Since `@elastic/eui` already ships with a module declaration, any local additions must be performed using module augmentation, e.g.

```typescript
import { CommonProps } from '@elastic/eui';
import { SFC } from 'react';

declare module '@elastic/eui' {
  export type EuiNewComponentProps = CommonProps & {
    additionalProp: string;
  };
  export const EuiNewComponent: SFC<EuiNewComponentProps>;
}
```

Reply:
I have moved the missing declaration for eui in a single file: https://github.com/elastic/kibana/pull/23950
It might be interesting to update the doc about it.

@weltenwort could you have a look as in this index.d.ts module augmentation there is no top level import or export.
cheers!

Reply:
that indeed seems to contradict my description :thinking: not sure what's going on here

Reply:
I added @weltenwort's feedback.  @sebelga - once your PR is merged (assuming it works... sounds like it might not?) we can update the documentation.